ocaml/testsuite/tests/typing-modes/portable-contend.ml

```diff
@@ -0,0 +1,249 @@
+(* TEST
+   * expect *)
+
+type r = {mutable a : string; b : string}
+
+(* TESTING records *)
+
+(* Reading/writing mutable field from contended record is rejected. Also note
+    that the mutation error precedes type error. *)
+let foo (r @ contended) = r.a <- 42
+[%%expect{|
+type r = { mutable a : string; b : string; }
+Line 7, characters 26-27:
+7 | let foo (r @ contended) = r.a <- 42
+                              ^
+Error: Found a contended value where a uncontended value was expected
+|}]
+
+let foo (r @ contended) = r.a
+[%%expect{|
+Line 1, characters 26-27:
+1 | let foo (r @ contended) = r.a
+                              ^
+Error: Found a contended value where a uncontended value was expected
+|}]
+
+(* reading immutable field from contended record is fine *)
+let foo (r @ contended) = r.b
+[%%expect{|
+val foo : r @ contended -> string @ contended = <fun>
+|}]
+
+(* Force top level to be uncontended and nonportable *)
+let r @ contended = "hello"
+[%%expect{|
+Line 1, characters 4-27:
+1 | let r @ contended = "hello"
+        ^^^^^^^^^^^^^^^^^^^^^^^
+Error: Found a contended value where a uncontended value was expected
+|}]
+
+let x @ portable = "world"
+
+let y @ portable = x
+[%%expect{|
+val x : string = "world"
+Line 3, characters 19-20:
+3 | let y @ portable = x
+                       ^
+Error: Found a nonportable value where a portable value was expected
+|}]
+
+(* Closing over writing mutable field gives nonportable *)
+let foo () =
+    let r = {a = "foo"; b = "bar"} in
+    let bar () = r.a <- "hello" in
+    let _ @ portable = bar in
+    ()
+[%%expect{|
+Line 4, characters 23-26:
+4 |     let _ @ portable = bar in
+                           ^^^
+Error: Found a nonportable value where a portable value was expected
+|}]
+
+(* Closing over reading mutable field gives nonportable *)
+let foo () =
+    let r = {a = "foo"; b = "bar"} in
+    let bar () = let _ = r.a in () in
+    let _ @ portable = bar in
+    ()
+[%%expect{|
+Line 4, characters 23-26:
+4 |     let _ @ portable = bar in
+                           ^^^
+Error: Found a nonportable value where a portable value was expected
+|}]
+
+(* Closing over reading immutable field is OK *)
+let foo () =
+    let r @ portable = {a = "foo"; b = "bar"} in
+    let bar () = let _ = r.b in () in
+    let _ @ portable = bar in
+    ()
+(* CR zqian: currently mutable(legacy) means all records constructed are nonportable,
+   and the above bar is closing over an nonportable record. Once we allow mutable()
+   syntax, we can test this. *)
+[%%expect{|
+Line 2, characters 23-45:
+2 |     let r @ portable = {a = "foo"; b = "bar"} in
+                           ^^^^^^^^^^^^^^^^^^^^^^
+Error: Found a nonportable value where a portable value was expected
+|}]
+
+
+(* TESTING arrays *)
+(* reading/writing to array requires uncontended *)
```

Comment:
It is boring now indeed as you said. But it will become less boring in the future as we mode cross most data types on `contention`, and these tests ensures that we haven't done too much to the point of unsoundness.